File: apps/quad/src/modules/processing/stages/timeline-merger.ts
/**
 * Merge per-player transcripts into a unified chronological timeline.
 *
 * The timeline is the central data format — all analysis reads from it.
 * Pure data transformations, no I/O.
 */

import type {
  TranscriptSegment,
  TimelineEntry,
  OverlapEvent,
  PlayerStats,
  CommunicationStats,
} from '../types.js';
import { logger } from '../../../core/logger.js';

/**
 * Merge per-player transcripts into a single chronological timeline.
 *
 * @param transcripts - Map of player name to their transcript segments
 * @returns Unified timeline sorted by start time, then speaker name for stable ordering
 */
export function mergeTranscripts(
  transcripts: Record<string, TranscriptSegment[]>,
): TimelineEntry[] {
  const timeline: TimelineEntry[] = [];

  for (const [playerName, segments] of Object.entries(transcripts)) {
    for (const seg of segments) {
      timeline.push({
        speaker: playerName,
        start: seg.start,
        end: seg.end,
        text: seg.text,
        confidence: seg.confidence ?? null,
      });
    }
  }

  timeline.sort((a, b) => a.start - b.start || a.speaker.localeCompare(b.speaker));

  logger.info('Merged transcripts into timeline', {
    players: Object.keys(transcripts).length,
    entries: timeline.length,
  });

  return timeline;
}

/**
 * Detect overlapping speech between different speakers.
 *
 * Uses the sorted order of the timeline to efficiently find overlaps:
 * for each entry, only check subsequent entries until their start >= current end.
 * Ignores same-speaker overlaps and tiny overlaps < 100ms.
 */
export function detectOverlaps(timeline: TimelineEntry[]): OverlapEvent[] {
  const overlaps: OverlapEvent[] = [];

  for (let i = 0; i < timeline.length; i++) {
    const entryA = timeline[i];

    for (let j = i + 1; j < timeline.length; j++) {
      const entryB = timeline[j];

      // Stop checking once we're past possible overlap
      if (entryB.start >= entryA.end) {
        break;
      }

      // Skip same-speaker overlaps
      if (entryA.speaker === entryB.speaker) {
        continue;
      }

      const overlapStart = Math.max(entryA.start, entryB.start);
      const overlapEnd = Math.min(entryA.end, entryB.end);
      const overlapDuration = overlapEnd - overlapStart;

      if (overlapDuration > 0.1) {
        const speakers = [entryA.speaker, entryB.speaker].sort() as [string, string];
        overlaps.push({
          speakers,
          start: Math.round(overlapStart * 1000) / 1000,
          end: Math.round(overlapEnd * 1000) / 1000,
          duration: Math.round(overlapDuration * 1000) / 1000,
          texts: {
            [entryA.speaker]: entryA.text,
            [entryB.speaker]: entryB.text,
          },
        });
      }
    }
  }

  logger.info('Detected speech overlaps', { count: overlaps.length });

  return overlaps;
}

/**
 * Compute per-player and team-wide communication statistics.
 *
 * @param timeline - Merged timeline entries
 * @param totalDuration - Optional total segment duration in seconds (for silence percentage)
 */
export function computeStats(
  timeline: TimelineEntry[],
  totalDuration?: number,
): CommunicationStats {
  const speakers: Record<string, { segments: number; totalSpeakingTime: number; totalWords: number }> = {};

  for (const entry of timeline) {
    const name = entry.speaker;
    if (!speakers[name]) {
      speakers[name] = { segments: 0, totalSpeakingTime: 0, totalWords: 0 };
    }
    const s = speakers[name];
    s.segments += 1;
    s.totalSpeakingTime += entry.end - entry.start;
    s.totalWords += entry.text.split(/\s+/).filter(w => w.length > 0).length;
  }

  const playerStats: Record<string, PlayerStats> = {};
  let totalSpeaking = 0;
  let totalSegments = 0;

  for (const [name, data] of Object.entries(speakers)) {
    const avgDuration = data.segments > 0
      ? Math.round((data.totalSpeakingTime / data.segments) * 100) / 100
      : 0;
    const wpm = data.totalSpeakingTime > 0
      ? Math.round((data.totalWords / (data.totalSpeakingTime / 60)) * 10) / 10
      : 0;

    playerStats[name] = {
      segments: data.segments,
      totalSpeakingTime: Math.round(data.totalSpeakingTime * 100) / 100,
      totalWords: data.totalWords,
      avgSegmentDuration: avgDuration,
      wordsPerMinute: wpm,
    };

    totalSpeaking += data.totalSpeakingTime;
    totalSegments += data.segments;
  }

  const playerCount = Object.keys(speakers).length;

  const stats: CommunicationStats = {
    playerStats,
    team: {
      totalSegments,
      totalSpeakingTime: Math.round(totalSpeaking * 100) / 100,
      playerCount,
    },
  };

  if (totalDuration !== undefined && totalDuration > 0) {
    stats.team.totalDuration = totalDuration;
    stats.team.silencePercentage = Math.round(
      (1 - totalSpeaking / (totalDuration * playerCount)) * 1000,
    ) / 10;
  }

  return stats;
}

/**
 * Format timeline entries as readable text for Claude analysis prompts.
 *
 * @param timeline - Merged timeline entries
 * @param maxEntries - Maximum number of entries to include (default 500)
 * @returns Formatted text like `[MM:SS.ss] Speaker: text`
 */
export function formatTimeline(timeline: TimelineEntry[], maxEntries = 500): string {
  const entries = timeline.slice(0, maxEntries);
  const lines: string[] = [];

  for (const entry of entries) {
    const totalSeconds = entry.start;
    const minutes = Math.floor(totalSeconds / 60);
    const seconds = totalSeconds - minutes * 60;
    const mm = String(minutes).padStart(2, '0');
    const ss = seconds.toFixed(2).padStart(5, '0');
    lines.push(`[${mm}:${ss}] ${entry.speaker}: ${entry.text}`);
  }

  if (timeline.length > maxEntries) {
    lines.push(`\n... (${timeline.length - maxEntries} more entries truncated)`);
  }

  return lines.join('\n');
}

File: archive/voice-analysis/src/processing/timeline_merger.py
"""Merge per-player transcripts into a unified chronological timeline.

The timeline is the central data format - all analysis reads from it.

Usage:
    python src/processing/timeline_merger.py <transcripts_dir>
"""

import sys
import json
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent.parent))
from src.utils.audio_utils import ensure_dir


def merge_transcripts(transcripts: dict[str, list[dict]]) -> list[dict]:
    """Merge per-player transcripts into chronological timeline.

    Args:
        transcripts: Dict mapping player name to transcript segments.

    Returns:
        Unified timeline sorted by start time:
        [
            {"speaker": "ParadokS", "start": 1.2, "end": 2.8, "text": "quad soon", ...},
            {"speaker": "Razor", "start": 1.5, "end": 3.1, "text": "im dead", ...},
        ]
    """
    timeline = []

    for player_name, segments in transcripts.items():
        for seg in segments:
            entry = {
                "speaker": player_name,
                "start": seg["start"],
                "end": seg["end"],
                "text": seg["text"],
                "confidence": seg.get("confidence"),
            }
            timeline.append(entry)

    # Sort by start time, then by speaker name for stable ordering
    timeline.sort(key=lambda e: (e["start"], e["speaker"]))

    return timeline


def detect_overlaps(timeline: list[dict]) -> list[dict]:
    """Detect overlapping speech between different speakers.

    Returns list of overlap events for analysis.
    """
    overlaps = []

    for i, entry_a in enumerate(timeline):
        for j in range(i + 1, len(timeline)):
            entry_b = timeline[j]

            # Stop checking once we're past possible overlap
            if entry_b["start"] >= entry_a["end"]:
                break

            if entry_a["speaker"] == entry_b["speaker"]:
                continue

            overlap_start = max(entry_a["start"], entry_b["start"])
            overlap_end = min(entry_a["end"], entry_b["end"])
            overlap_duration = overlap_end - overlap_start

            if overlap_duration > 0.1:  # Ignore tiny overlaps < 100ms
                overlaps.append({
                    "speakers": sorted([entry_a["speaker"], entry_b["speaker"]]),
                    "start": round(overlap_start, 3),
                    "end": round(overlap_end, 3),
                    "duration": round(overlap_duration, 3),
                    "texts": {
                        entry_a["speaker"]: entry_a["text"],
                        entry_b["speaker"]: entry_b["text"],
                    },
                })

    return overlaps


def compute_stats(timeline: list[dict], total_duration: float | None = None) -> dict:
    """Compute per-player communication statistics."""
    speakers = {}

    for entry in timeline:
        name = entry["speaker"]
        if name not in speakers:
            speakers[name] = {
                "segments": 0,
                "total_speaking_time": 0.0,
                "total_words": 0,
            }

        s = speakers[name]
        s["segments"] += 1
        s["total_speaking_time"] += entry["end"] - entry["start"]
        s["total_words"] += len(entry["text"].split())

    player_stats = {}
    total_speaking = 0.0
    total_segments = 0

    for name, data in speakers.items():
        avg_length = data["total_speaking_time"] / data["segments"] if data["segments"] else 0
        player_stats[name] = {
            "segments": data["segments"],
            "total_speaking_time": round(data["total_speaking_time"], 2),
            "total_words": data["total_words"],
            "avg_segment_duration": round(avg_length, 2),
            "words_per_minute": round(data["total_words"] / (data["total_speaking_time"] / 60), 1)
            if data["total_speaking_time"] > 0 else 0,
        }
        total_speaking += data["total_speaking_time"]
        total_segments += data["segments"]

    stats = {
        "player_stats": player_stats,
        "team": {
            "total_segments": total_segments,
            "total_speaking_time": round(total_speaking, 2),
            "player_count": len(speakers),
        },
    }

    if total_duration:
        stats["team"]["total_duration"] = total_duration
        stats["team"]["silence_percentage"] = round(
            (1 - total_speaking / (total_duration * len(speakers))) * 100, 1
        )

    return stats


def main():
    """CLI entrypoint."""
    if len(sys.argv) < 2:
        print("Usage: python timeline_merger.py <transcripts_dir>")
        sys.exit(1)

    transcripts_dir = Path(sys.argv[1])

    # Load per-player transcripts
    transcripts = {}
    for json_file in sorted(transcripts_dir.glob("*.json")):
        if json_file.name.startswith("merged_") or json_file.name.startswith("stats") or json_file.name == "overlaps.json":
            continue
        player_name = json_file.stem
        with open(json_file) as f:
            transcripts[player_name] = json.load(f)

    if not transcripts:
        print(f"No transcript files found in {transcripts_dir}")
        sys.exit(1)

    timeline = merge_transcripts(transcripts)
    overlaps = detect_overlaps(timeline)
    stats = compute_stats(timeline)

    # Write outputs
    out_timeline = transcripts_dir / "merged_timeline.json"
    with open(out_timeline, "w") as f:
        json.dump(timeline, f, indent=2)
    print(f"Timeline: {out_timeline} ({len(timeline)} entries)")

    out_overlaps = transcripts_dir / "overlaps.json"
    with open(out_overlaps, "w") as f:
        json.dump(overlaps, f, indent=2)
    print(f"Overlaps: {out_overlaps} ({len(overlaps)} events)")

    out_stats = transcripts_dir / "stats.json"
    with open(out_stats, "w") as f:
        json.dump(stats, f, indent=2)
    print(f"Stats: {out_stats}")

    # Print summary
    print(f"\n{'Player':<15s} {'Segs':>5s} {'Time':>7s} {'Words':>6s} {'WPM':>5s}")
    print("-" * 40)
    for name, ps in stats["player_stats"].items():
        print(f"{name:<15s} {ps['segments']:>5d} {ps['total_speaking_time']:>6.1f}s {ps['total_words']:>6d} {ps['words_per_minute']:>5.1f}")


if __name__ == "__main__":
    main()
